### PR TITLE
[3.2] Lightness of dynamic objects from Energy BakedLightmapData

### PR DESCRIPTION
**Pull request by dedm0zaj**
Added the effect of Energy BakedLightmapData on dynamic objects.

image
image

On the GIF: cubes are dynamic objects.

gd_capture_energy

Works well in a running game.
In the editor, for the changes to the Energy parameter to take effect, you need to turn off and on the visibility of BakedLightmap.

image

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/1952.

### CODE REVIEW

**Review by clayjohn — COMMENTED** [review]
Looks reasonable to me. But I am unfamiliar with this code. Will need to wait for @JFonS to approve.

**Comment by JFonS** [comment]
With this approach we need to make sure `_update_instance_lightmap_captures` is called whenever the energy is changed. Otherwise we can have instances that don't update properly.

I think a better alternative is to multiply the lightmap energy directly in the shader. That would be in: 
https://github.com/godotengine/godot/blob/cddff0404dbeadbe04d575c0211adba7fa4b3355/drivers/gles3/shaders/scene.glsl#L1856-L1861
and for GLES2:
https://github.com/godotengine/godot/blob/cddff0404dbeadbe04d575c0211adba7fa4b3355/drivers/gles3/shaders/scene.glsl#L1856-L1861

The `lightmap_energy` is already available, so there should be no update issues.

**Comment by dedm0zaj** [comment]
@JFonS 
I tried to do through shaders, but i am having problems:
1. In scene.glsl, the `lightmap_energy` variable refers to the `#ifdef USE_LIGHTMAP` section.
And it must be applied in the `#ifdef USE_LIGHTMAP_CAPTURE` section.
I didn't understand how to do it.
2. I created my `capture_energy` variable for the `#ifdef USE_LIGHTMAP_CAPTURE` section. And I tried to convey its meaning here:
https://github.com/godotengine/godot/blob/cddff0404dbeadbe04d575c0211adba7fa4b3355/drivers/gles3/rasterizer_scene_gles3.cpp#L1941-L1945
For this, I added a few lines:
```
RasterizerStorageGLES3::LightmapCapture *capture = storage->lightmap_capture_data_owner.getornull(e->instance->lightmap_capture->base);
state.scene_shader.set_uniform(SceneShaderGLES3::CAPTURE_ENERGY, capture->energy);
```
but `e->instance->lightmap_capture` is NULL

I still haven't found a way to do it through the shader.

**Comment by JFonS** [comment]
ah right, my bad. I haven't looked deeply into it, but it seems like it won't be easy to get the capture energy when setting up the shader uniforms. So we're back to your implementation :)

I say we merge this PR and, if we find any case where capture is not updated after changing the energy, we can look into it.

**Review by JFonS — APPROVED** [review]

**Comment by dedm0zaj** [comment]
@JFonS
There is an update problem in the editor. To update, you need to turn off and on the visibility of BakedLightmap.
I mentioned this in the topic.



If the change is done in the shader, then this problem will not occur.

**Comment by dedm0zaj** [comment]
@JFonS oops! I accidentally pressed "Re-request review". I don't know much about github and its functions yet. Sorry

**Comment by JFonS** [comment]
@dedm0zaj No worries :)

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**
